fem/kernels/kIntDiffusionMultAdd.cpp

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2017, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-734707. All Rights
+// reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include "../../general/okina.hpp"
+
+// *****************************************************************************
+#define QUAD_2D_ID(X, Y) (X + ((Y) * NUM_QUAD_1D))
+
+// *****************************************************************************
+template<const int NUM_DOFS_1D,
+         const int NUM_QUAD_1D>
+void kIntDiffusionMultAdd2D(const int numElements,
+                            const double* __restrict dofToQuad,
+                            const double* __restrict dofToQuadD,
+                            const double* __restrict quadToDof,
+                            const double* __restrict quadToDofD,
+                            const double* __restrict oper,
+                            const double* __restrict solIn,
+                            double* __restrict solOut)
+{
+   forall(e, numElements,
```

Comment:
How do you use shared memory with your `forall`?

Reply:
We cannot just with the `forall`, the `shared` local arrays will be usable with some codegen + JIT features.